software/base/src/main/java/org/apache/brooklyn/entity/software/base/AbstractSoftwareProcessSshDriver.java

```diff
@@ -344,9 +346,23 @@ public void runPostLaunchCommand() {
      * @see SoftwareProcess#SHELL_ENVIRONMENT
      */
     public Map<String, String> getShellEnvironment() {
-        return Strings.toStringMap(entity.getConfig(SoftwareProcess.SHELL_ENVIRONMENT), "");
+        Map<String, Object> env = entity.getConfig(SoftwareProcess.SHELL_ENVIRONMENT);
+        if (env == null) {
+            return null;
+        }
+        ShellEnvironmentSerializer envSerializer = new ShellEnvironmentSerializer(((EntityInternal)entity).getManagementContext());
+        Map<String, String> serializedEnv = Maps.newHashMap();
+        for (Entry<String, Object> entry : env.entrySet()) {
+            String key = serializeShellEnv(envSerializer, entry.getKey());
```

Comment:
That's to keep backwards compatibility. `Strings.toStringMap` would call `toString` on the keys.
Using invalid keys will fail the task on the bash side. I'd like to avoid validating the key here as depending on where you look you'll find different definitions for a valid name.